### ISSUE TITLE
Can't download vods anymore (bug)

### ISSUE THREAD

**Issue by H-F-T** [ihabunek/twitch-dl#35]
It seems that twitch has changed/updated their m3u8 file handling and now downloading vod doesn't work property anymore.

**Comment by ihabunek** [comment]
Of course they have :) I'll look into it.

**Comment by ihabunek** [comment]
Hm, I've tried downloading one video and that worked:

```
twitch-dl download 715300612 --end 0:01
```

 Can you give me a command which fails for you?

**Comment by H-F-T** [comment]
The download works but if you check the file it downloads too many chunks. Download 3 min of it and video turns out 20 mins long constantly looping few seconds back. Try with this vod 728359768

**Comment by H-F-T** [comment]
TwitchLeecher seem to have resolved it [https://github.com/Franiac/TwitchLeecher/pull/515](https://github.com/Franiac/TwitchLeecher/pull/515)
New format contains up to 5 duplicates for each chunk.

**Comment by bouffdaddy** [comment]
Investigating offsets, they have also extended chunk length to 9 seconds

**Comment by kadedworkin** [comment]
Seems to only have affected videos created in a certain time frame.  Those videos effected have playlist.m3u8 files with this 2 second interval declared and inappropriately repeated.

'#EXTINF:2.000,
#EXT-X-BYTERANGE:1090588@0
0.ts
#EXTINF:2.000,
#EXT-X-BYTERANGE:1086076@1090588
0.ts
#EXTINF:2.000,
#EXT-X-BYTERANGE:1087016@2176664
0.ts
#EXTINF:2.000,
#EXT-X-BYTERANGE:1088896@3263680
0.ts
#EXTINF:2.000,
#EXT-X-BYTERANGE:1089836@4352576
0.ts'

While working to find a solution to handle this new m3u8 format, I pulled a recently completed stream and was pleasantly surprised with the expected m3u8 with 10 second data. 

`#EXTINF:10.000,
0.ts
#EXTINF:10.000,
1.ts
#EXTINF:10.000,
2.ts
#EXTINF:10.000,
3.ts
#EXTINF:10.000,
4.ts
#EXTINF:10.000,
5.ts`

Sure looks like they rolled back some kind of change server side. Accordingly, downloader now works as expected. Those files with the bad m3u8s remain bad though.

**Comment by garoto** [comment]
Just chiming in to say that ffmpeg apparently supports this `#EXT-X-BYTERANGE:` format out-of-the-box since, welp, a reasonable long time ago I guess, because my `mpv` build from January this year can still play the examples I was given on IRC just fine.

Example:

```
ffmpeg/demuxer: hls: HLS request for url 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts', offset 0, playlist 0
    ffmpeg: https: Opening 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts' for reading
ffmpeg/demuxer: hls: HLS request for url 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts', offset 709888, playlist 0
    ffmpeg: https: Opening 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts' for reading
ffmpeg/demuxer: hls: HLS request for url 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts', offset 1407932, playlist 0
    ffmpeg: https: Opening 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts' for reading
ffmpeg/demuxer: hls: HLS request for url 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts', offset 2123460, playlist 0
    ffmpeg: https: Opening 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts' for reading
ffmpeg/demuxer: hls: HLS request for url 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts', offset 2814172, playlist 0
    ffmpeg: https: Opening 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/149.ts' for reading
ffmpeg/demuxer: hls: HLS request for url 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/150.ts', offset 0, playlist 0
    ffmpeg: https: Opening 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/150.ts' for reading
ffmpeg/demuxer: hls: HLS request for url 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/150.ts', offset 711956, playlist 0
    ffmpeg: https: Opening 'https://d2nvs31859zcd8.cloudfront.net/56b0bf07ebd4e1d7d06a_emperorblackman_39514584382_1598986657/chunked/150.ts' for reading
```

**Comment by ihabunek** [comment]
Thanks for the info everyone.

The strategy which seems like it should work, instead of just joining vods sequentially:

- download all vods from twitch
- rewrite playlist to reference the downloaded vods
- give the playlist as input to ffmpeg

**Comment by H-F-T** [comment]
[Looks like Twitch has reverted the change for now and newly made vods work yet again](https://www.reddit.com/r/Twitch/comments/ileilm/cant_download_certain_vod/g3seer4/). The vods made in time period of that change seem to be broken for good and they can be used to test your new approach.
Like usual it's not clear what Twitch is doing and if they are fixing or reverting back and it could be that they will come back with new format again.

**Comment by ihabunek** [comment]
@H-F-T @garoto @kadedworkin I made a fix as described above in bf2a455, I'd love if you could test it for me.

I'm also testing a new way to distribute twitch-dl as standalone executable which only require python3. :)

To test:
* download this file: http://bezdomni.net/twitch-dl.pyz
* to check there's no funny business, unzip the file, it contains the twitch-dl source code with all dependencies bundled
* then either:
    * make it executable and run it `./twitch-dl.pyz` (linux specific)
    * pass it to python: `python3 twitch-dl.pyz`

Now you can use it instead of your installed version of twitch-dl, e.g. 

```
python3 twitch-dl.pyz download 728359768 --quality 160p
```

I've tested it on `728359768`, as suggested by @H-F-T but I'd appreciate you try it out a bit and tell me if you run into any issues. If all is fine, I'll make a release.

Thanks!

**Comment by ihabunek** [comment]
@H-F-T Good catch, can reproduce, looking into it.

**Comment by H-F-T** [comment]
In addition I just ran your given example `python twitch-dl.pyz download 728359768 --quality 160p ` to see if result is different.
```
Joining files...
[2mffmpeg -i C:\Users\stati\AppData\Local\Temp/twitch-dl/f603e68fb3f52bc60f50_jinnytty_39515155374_1598990431/160p30/playlist_downloaded.m3u8 -c copy 20200901_728359768_jinnytty_hahehehaheheh.mkv -stats -loglevel warning[0m
[hls @ 06763b40] Failed to open segment 0 of playlist 0
[hls @ 06763b40] Failed to open segment 1 of playlist 0
[hls @ 06763b40] Failed to open segment 2 of playlist 0
...cut out the middle for less clutter as it just counts from 0 to 7725...
[hls @ 06763b40] Failed to open segment 7724 of playlist 0
[hls @ 06763b40] Failed to open segment 7725 of playlist 0
[hls @ 06763b40] Error when loading first segment 'C:\Users\stati\AppData\Local\Temp/twitch-dl/f603e68fb3f52bc60f50_jinnytty_39515155374_1598990431/160p30/C:\Users\stati\AppData\Local\Temp/twitch-dl/f603e68fb3f52bc60f50_jinnytty_39515155374_1598990431/160p30/00000.ts'
C:\Users\stati\AppData\Local\Temp/twitch-dl/f603e68fb3f52bc60f50_jinnytty_39515155374_1598990431/160p30/playlist_downloaded.m3u8: Invalid data found when processing input
[91mJoining files failed[0m
```

**Comment by ihabunek** [comment]
@H-F-T Can you attach the playlist from `C:\Users\stati\AppData\Local\Temp/twitch-dl/f603e68fb3f52bc60f50_jinnytty_39515155374_1598990431/160p30/playlist_downloaded.m3u8`? Just rename it to .txt so github allows it.

It's probably a windows issue.

**Comment by H-F-T** [comment]
[playlist_downloaded.m3u8.txt](https://github.com/ihabunek/twitch-dl/files/5167693/playlist_downloaded.m3u8.txt)

**Comment by ihabunek** [comment]
Here's a modified playlist which might fix the issue (replaced / with \ for path separators): [playlist.m3u8.txt](https://github.com/ihabunek/twitch-dl/files/5167723/playlist.m3u8.txt)

Please download it, rename to `playlist.m3u8`, and run:

```
ffmpeg -i playlist.m3u8 -c copy out.mkv -stats -loglevel warning
```

I'm presuming the VOD files are still in the temp dir since you used `--keep`.

**Comment by H-F-T** [comment]
Yes that worked.

**Comment by ihabunek** [comment]
@H-F-T I believe I fixed both issues. Here's a new version: http://bezdomni.net/twitch-dl.pyz

Please test again.

**Comment by H-F-T** [comment]
Everything seems to be working now, tested with multiple vods.
Thanks for fixing and improving it.

**Comment by ihabunek** [comment]
Released [1.11.0](https://github.com/ihabunek/twitch-dl/releases/tag/1.11.0) with these fixes.

If you have a minute, check out the [new installation instructions ](https://github.com/ihabunek/twitch-dl#installation) and let me know if they make sense. Tried to make it simpler for less experienced users.
